tests/test_data_file.py

```diff
@@ -103,3 +103,133 @@ def test_data_file_all(snap, atom_style, set_style, shuffle_ids, tmp_path):
     else:
         assert not snap_2.has_charge()
+
+
+@pytest.mark.parametrize("shuffle_ids", [False, True])
+def test_data_file_topology(snap_top, tmp_path, shuffle_ids):
+    # set ids to be assigned
+    if shuffle_ids:
+        particle_id = [1, 5, 2, 6, 3, 7, 4, 8]
+        bond_id = [1, 4, 2, 5, 3, 6]
+        angle_id = [1, 3, 2, 4]
+        dihedral_id = [2, 1]
+        improper_id = [2, 1]
+    else:
+        particle_id = [1, 2, 3, 4, 5, 6, 7, 8]
+        bond_id = [1, 2, 3, 4, 5, 6]
+        angle_id = [1, 2, 3, 4]
+        dihedral_id = [1, 2]
+        improper_id = [1, 2]
+
+    # particle information
+    snap_top.id = particle_id
+    snap_top.typeid = [1, 1, 1, 1, 2, 2, 2, 2]
+    snap_top.position = [
+        [0, 0, 0],
+        [0.1, 0.1, 0.1],
+        [0.2, 0.2, 0.2],
+        [0.3, 0.3, 0.3],
+        [1, 1, 1],
+        [1.1, 1.1, 1.1],
+        [1.2, 1.2, 1.2],
+        [1.3, 1.3, 1.3],
+    ]
+    snap_top.mass = [1, 1, 1, 1, 2, 2, 2, 2]
+
+    # bond information
+    snap_top.bonds = lammpsio.topology.Bonds(N=6, num_types=2)
+    snap_top.bonds.id = bond_id
+    snap_top.bonds.typeid = [1, 2, 1, 2, 1, 2]
+    snap_top.bonds.members = [
+        [1, 2],
+        [2, 3],
+        [3, 4],
+        [5, 6],
+        [6, 7],
+        [7, 8],
+    ]
+
+    # angle information
+    snap_top.angles = lammpsio.topology.Angles(N=4, num_types=2)
+    snap_top.angles.id = angle_id
+    snap_top.angles.typeid = [1, 2, 2, 1]
+    snap_top.angles.members = [
+        [1, 2, 3],
+        [2, 3, 4],
+        [5, 6, 7],
+        [6, 7, 8],
+    ]
+
+    # dihedral information
+    snap_top.dihedrals = lammpsio.topology.Dihedrals(N=2, num_types=2)
+    snap_top.dihedrals.id = dihedral_id
+    snap_top.dihedrals.typeid = [1, 2]
+    snap_top.dihedrals.members = [
+        [1, 2, 3, 4],
+        [5, 6, 7, 8],
+    ]
+
+    # improper information
+    snap_top.impropers = lammpsio.topology.Impropers(N=2, num_types=2)
+    snap_top.impropers.id = improper_id
+    snap_top.impropers.typeid = [1, 2]
+    snap_top.impropers.members = [
+        [1, 2, 3, 4],
+        [5, 6, 7, 8],
+    ]
+
+    filename = tmp_path / "atoms.data"
+    data = lammpsio.DataFile.create(filename, snap_top)
+    assert filename.exists
+
+    snap_2 = data.read()
+    # test none topology features of a snapshot with topology
+    assert snap_2.N == snap_top.N
+    if shuffle_ids:
+        assert snap_2.has_id()
+        assert numpy.allclose(snap_2.id, snap_top.id)
+    else:
+        assert not snap_2.has_id()
+    assert numpy.allclose(snap_2.id, snap_top.id)
+    assert snap_2.has_typeid()
+    assert numpy.allclose(snap_2.typeid, snap_top.typeid)
+    assert snap_2.has_position()
+    assert numpy.allclose(snap_2.position, snap_top.position)
+    assert snap_2.has_mass()
+    assert numpy.allclose(snap_2.mass, snap_top.mass)
+
+    # test bonds
+    assert snap_2.bonds.N == snap_top.bonds.N
+    assert snap_2.bonds.has_id()
```

Comment:
What is the purpose of not saving the id if they are in order? Should that be the intended behavior for the topology ids as well?

Reply:
I think the idea was that by default in the `Snapshot`, someone may not assign the ID if they don't care / if the particles are already in order (and so their index in the NumPy arrays is the same as their ID, offset by 1). The reasoning was, in that case, why bother saving the values, if they can be deduced?

It is also possible for IDs to have gaps in them, as far as I remember (i.e., they don't have to be a compact set [1,N]). This could happen if someone deleted one, for example, which I think LAMMPS supports.

I think we should probably do this for the topologies as well. Fortunately, that is simpler than for the `Snapshot` because the ID is only used in one section, and we don't have to make sure everything is sorted consistently across sections. (Maybe LAMMPS actually guarantees that for particles, but I wasn't sure/didn't trust it haha.) Basically, you would do a check like:
```
# only write ID if it is out of default order
if id != i + 1:
    # write the ID value for index i
```